src/cli/AttachmentExport.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ *  Copyright (C) 2019 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 2 or (at your option)
+ *  version 3 of the License.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <cstdio>
+#include <cstdlib>
+
+#include "AttachmentExport.h"
+
+#include "core/Group.h"
+
+#include <QFile>
+
+AttachmentExport::AttachmentExport(FILE* fout) : m_fout(fout)
+{
+    name = QString("attachment-export");
+    description = QObject::tr("Export an attachment of an entry.");
+    positionalArguments.append(
+        {QString("entry"), QObject::tr("Path of the entry with the target attachment."), QString("")});
+    positionalArguments.append(
+        {QString("name"), QObject::tr("Name of the attachment to be exported."), QString("")});
+}
+
+int AttachmentExport::executeWithDatabase(QSharedPointer<Database> database, QSharedPointer<QCommandLineParser> parser)
+{
+    auto& err = Utils::STDERR;
+
+    const QStringList args = parser->positionalArguments();
```

Comment:
Hi ! (I'm not sure if I'm allowed to review), however I want to point out will have to null/sanity check your `database` & `parser` objects right after passing them to your functions or you will get segmentation-faults in case of misinitialization.

Reply:
Hey there! That would surely be the most defensive approach to it but none of the other commands I have checked do those checks which is why I figured it would not be needed. For now I'll wait and see what others have to say about it.

Reply:
Just because in other commands the asserts/checks were omitted it doesn't necessarily mean that this is the right way to do it. (I'd suggest fixing them aswell)

What would you suppose should happen in a case of a null object ? You then have undefined behaviour and not handeld an error case properly. Another option could be to pass the function arguments as references.

Reply:
But a null database or parser object would never reach an implementation of `DatabaseCommand::executeWithDatabase` because `DatabaseCommand::execute`, which later invokes that method, already does the null checking for us [here](https://github.com/keepassxreboot/keepassxc/blob/fd3cc7e8c3d012d74d649773eb2236bd06f12fbc/src/cli/DatabaseCommand.cpp#L40-L42) and [here](https://github.com/keepassxreboot/keepassxc/blob/fd3cc7e8c3d012d74d649773eb2236bd06f12fbc/src/cli/DatabaseCommand.cpp#L61-L63), so we would only be duplicating code.

Reply:
> 
> 
> But a null database or parser object would never reach an implementation of `DatabaseCommand::executeWithDatabase` because `DatabaseCommand::execute`, which later invokes that method, already does the null checking for us [here](https://github.com/keepassxreboot/keepassxc/blob/fd3cc7e8c3d012d74d649773eb2236bd06f12fbc/src/cli/DatabaseCommand.cpp#L40-L42) and [here](https://github.com/keepassxreboot/keepassxc/blob/fd3cc7e8c3d012d74d649773eb2236bd06f12fbc/src/cli/DatabaseCommand.cpp#L61-L63), so we would only be duplicating code.

You are right in this specific case the caller has taken care of the null check, but IMO null checking should be in the callee (the called function) were the variable goes into scope. 
This defeats the case that if someone else calls your function & _does not null check before_ for whatever reason (intentionally or unintentionally) you'd drive into the same problem again. 

So for every function call of yours the caller would have to write a designated null check.

Reply:
> I'm not sure if I'm allowed to review

@Colfenor Of course you are!

Reply:
@Colfenor While I agree with you I think the changes you talk about should be part of a different pull request that fixes all existing DatabaseCommands. An alternative would be marking `executeWithDatabase` as protected to prevent invocation from outside classes.

Reply:
We'll if you agree, may I ask what is stopping you from fixing them in your PR ? (or at least sanitize the arguments (null-check) with your three new commands)

Reply:
Sorry for the delayed response. To answer your question: Because I really don't think of it as a change which is too important. It certainly doesn't have downsides to it but there is null safety in place already (albeit not directly inside the method) and I think it would make more sense for a single commit to fix all occurrences of it, which you could tackle once this pull request has been merged.
I'd prefer dropping the null debate for now if you don't mind but I would of course still be happy about additional feedback regarding the code I wrote.

Reply:
@andreblanke I agree that we can defer that specific (null checking) change to a later, dedicated PR. I should indeed already be done before calling `executeWithDatabase`.